design-documents/database/sequence-functionality-in-magento.md

```diff
@@ -0,0 +1,132 @@
+## Sequence functionality in Magento
+
+### Terms
+
+**Entity UUID** - universally unique identifier, user-defined OR auto-generated by system
+
+### Overview
+
+#### Sequence mechanism
+In Magento there is a sequence mechanism, which provides system-generated entity identifier. We already have this mechanism for the following entities:
+
+**Magento CE (SalesSequence):**
+* Order
+* Invoice
+* Shipment
+* Credit memo
+
+**Magento EE (Staging):**
+* Product
+* Category
+* CMS page
+* CMS block
+* Catalog rule
+* Sales rule
+* Product bundle option
+* Product bundle section
+
+As you can see in Magento EE we have much more entities that use sequence mechanism. 
+This causes bugs when CE specific functionality are not fully compatible with the EE, due to the fact that some entities use sequence mechanism in EE.
+This affects the contributions of community members.
+
+#### Staging specific database changes
+The following entities in EE even use the new primary key for main entity tables - **_row_id_**:
+* Product
+* Category
+* CMS page
+* CMS block
+* Catalog rule
+* Sales rule
+
+These changes were introduced by the Staging functionality. 
+The part of functionality that is responsible for updating of foreign keys is hidden in schema Recurring scripts not even in Staging modules:
+* app/code/Magento/AdvancedSalesRule/Setup/Recurring.php
+* app/code/Magento/Banner/Setup/Recurring.php
+* app/code/Magento/CatalogEvent/Setup/Recurring.php
+* app/code/Magento/CatalogPermissions/Setup/Recurring.php
+* app/code/Magento/GiftRegistry/Setup/Recurring.php
+* app/code/Magento/Reminder/Setup/Recurring.php
+* app/code/Magento/TargetRule/Setup/Recurring.php
+* app/code/Magento/VersionsCms/Setup/Recurring.php
+* app/code/Magento/VisualMerchandiser/Setup/Recurring.php
+
+The solution is implemented declaratively using DI, but it still causes some issues with upgrade in EE. 
+
+#### Declarative Schema issue
+Due to specific install/upgrade flow, when we use Declarative Schema, most of Staging foreign keys recreates on each upgrade.
+
+We have 3 ways to explicitly declare new Staging sequence foreign keys:
+
+1. Add 10 new dependencies into 4 Staging modules
```

Comment:
It would be helpful to cover what exact changes do we need in the modules, that require adding dependencies. Same for all solutions. It's not very clear what exactly should be done in the modules. Dependencies are just consequences.